### PR TITLE
Expiration times

### PR DESCRIPTION
**Pull request by acdlite**
An expiration time represents a time in the future by which an update should flush. The priority of the update is related to the difference between the current clock time and the expiration time. This has the effect of increasing the priority of updates as time progresses, to prevent starvation.

### CODE REVIEW

**Comment by acdlite** [comment]
Confirms that this makes the Triangle demo work as expected. Still need to write some unit tests, but this is ready for review.

**Comment by acdlite** [comment]
Unit tests are completed

File: src/renderers/testing/ReactTestRendererFiberEntry.js [hunk 1]
```diff
@@ -249,6 +249,11 @@ var TestRenderer = ReactFiberReconciler({
   useSyncScheduling: true,
 
   getPublicInstance,
+
+  now(): number {
+    // Test renderer does not use expiration
```
Comment (acdlite) [inline src/renderers/testing/ReactTestRendererFiberEntry.js]: @sebmarkbage I went back and forth on whether this should be an optional method. Landed on making it required because all renderers will need to account for features like expiration boundaries.

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js [hunk 2]
```diff
@@ -163,6 +163,22 @@ function shouldAutoFocusHostComponent(type: string, props: Props): boolean {
   return false;
 }
 
+// TODO: Better polyfill
```
Comment (acdlite) [inline src/renderers/dom/fiber/ReactDOMFiberEntry.js]: Is this polyfill sufficient? Or am I overlooking something?

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js [hunk 3]
```diff
@@ -163,6 +163,22 @@ function shouldAutoFocusHostComponent(type: string, props: Props): boolean {
...
+let now;
+if (
+  typeof window !== 'undefined' &&
+  window.performance &&
+  typeof window.performance.now === 'function'
+) {
+  now = function() {
```
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberEntry.js]: Why not `now = performance.now` and `now = Date.now`?

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js (hunk 2 above)
Comment (sebmarkbage) [inline src/renderers/dom/fiber/ReactDOMFiberEntry.js]: It's fine. Maybe we should colocate it with `ReactDOMFrameScheduling` to have all these in one place?

It is also `performance.now` aware: https://github.com/facebook/react/blob/master/src/renderers/shared/ReactDOMFrameScheduling.js#L73

File: src/renderers/dom/fiber/ReactDOMFiberEntry.js (hunk 3 above)
Comment (acdlite) [inline src/renderers/dom/fiber/ReactDOMFiberEntry.js]: `performance.now` needs to be bound, but looks like `Date.now` doesn't. I'll update.

File: src/renderers/shared/fiber/ReactFiberCompleteWork.js [hunk 4]
```diff
@@ -42,7 +42,7 @@ var {
   Fragment,
 } = ReactTypeOfWork;
 var {Placement, Ref, Update} = ReactTypeOfSideEffect;
-var {OffscreenPriority} = ReactPriorityLevel;
+var {Never} = ReactFiberExpirationTime;
```
Comment (flarnie) [inline src/renderers/shared/fiber/ReactFiberCompleteWork.js]: I like that this name change decouples this expiration time from the UI context in which it's used. What we really mean is 'this might never finish updating, and that's ok' and that may or may not be literally "offscreen" in the UI.

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+'use strict';
+
+import type {PriorityLevel} from 'ReactPriorityLevel';
+const {
+  NoWork,
+  SynchronousPriority,
+  TaskPriority,
+  HighPriority,
+  LowPriority,
+  OffscreenPriority,
```
Comment (flarnie) [inline src/renderers/shared/fiber/ReactFiberExpirationTime.js]: haha, I guess it's not a rename after all. ("OffscreenPriority" to "Never" expirationTime)

**Comment by acdlite** [comment]
@gaearon I’ve rebased this locally, just haven’t pushed yet. I’ll work on getting all my PRs updated tomorrow.

File: src/renderers/native/ReactNativeFiberRenderer.js [hunk 6]
```diff
@@ -377,6 +377,11 @@ const NativeRenderer = ReactFiberReconciler({
   },
 
   useSyncScheduling: true,
+
+  now(): number {
```
Comment (flarnie) [inline src/renderers/native/ReactNativeFiberRenderer.js]: Why is this added to `ReactNativeFiberRenderer` and not `ReactNativeFiberEntry`?

Comment (acdlite) [reply src/renderers/native/ReactNativeFiberRenderer.js]: "Entry" modules are the public API. The renderer is what you create out of a host config. For whatever reason, our React Native renderer splits the two into separate files, whereas our DOM renderer does not. Don't think there's any real reason, just happened to be factored that way.

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+const Done = 0;
+const Sync = 1;
+const Task = 2;
+const Never = Number.MAX_SAFE_INTEGER;
+
+const UNIT_SIZE = 10;
+const MAGIC_NUMBER_OFFSET = 10;
+
+exports.Done = Done;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberExpirationTime.js]: How about keeping the `NoWork` name here? I also wonder if the value should be named something that has something that indicates that it is really a dirty flag and that this essentially means false. `Done` stood out to me as not indicating that well unless you already know how it works.

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+const invariant = require('fbjs/lib/invariant');
+
+// TODO: Use an opaque type once ESLint et al support the syntax
+export type ExpirationTime = number;
+
+const Done = 0;
+const Sync = 1;
+const Task = 2;
+const Never = Number.MAX_SAFE_INTEGER;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberExpirationTime.js]: This number is way too high. Since VMs (really mostly just V8) tries to optimize for integer types when they can, sometimes they want them to fit into a single 32bit word on 32-bit systems. This number might cause them to deopt that assumption and have to expand the slot.

It should be 2^30 or whatever is the highest we can fit in 31-bit.

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+const UNIT_SIZE = 10;
+const MAGIC_NUMBER_OFFSET = 10;
+
+exports.Done = Done;
+exports.Never = Never;
+
+// 1 unit of expiration time represents 10ms.
+function msToExpirationTime(ms: number): ExpirationTime {
+  // Always add an offset so that we don't clash with the magic number for Done.
+  return Math.round(ms / UNIT_SIZE) + MAGIC_NUMBER_OFFSET;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberExpirationTime.js]: Why round? (Faster to do `(ms / UNIT_SIZE) | 0`.)

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+}
+exports.msToExpirationTime = msToExpirationTime;
+
+function ceiling(num: number, precision: number): number {
+  return Math.ceil(Math.ceil(num * precision) / precision);
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberExpirationTime.js]: Maybe `... | 0 + 1` would be faster and good enough?

File: src/renderers/shared/fiber/ReactFiberScheduler.js [hunk 11]
```diff
@@ -801,14 +820,12 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
               'a bug in React. Please file an issue.',
           );
           // We just completed a root. Commit it now.
-          priorityContext = TaskPriority;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberScheduler.js]: Why is this not needed anymore?

**Comment by sebmarkbage** [comment]
After the nits I think this is good to go.

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+          'This error is likely caused by a bug in React. Please file an issue.',
+      );
+  }
+}
+exports.priorityToExpirationTime = priorityToExpirationTime;
+
+// Given the current clock time and an expiration time, returns the
+// corresponding priority level. The more time has advanced, the higher the
+// priority level.
+function expirationTimeToPriorityLevel(
```
[the remark this answers is not in the document]
Comment (NE-SmallTown) [reply src/renderers/shared/fiber/ReactFiberExpirationTime.js]: What do you mean?We are based on "ReactPriorityLevel" to scheduling, right?

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js (hunk 7 above)
Comment (acdlite) [reply src/renderers/shared/fiber/ReactFiberExpirationTime.js]: Yeah I'm not happy with any of the names here, was just trying to avoid a naming clash. `NoWork` seems fine, can import the other one as `NoWorkPriority` if I need to.

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (acdlite) [reply src/renderers/shared/fiber/ReactFiberExpirationTime.js]: This function used after checking the deadline so we know whether to [resume or yield](https://github.com/acdlite/react/blob/867aeac5819c997bfee0a0ddf7d973624ef5e3d2/src/renderers/shared/fiber/ReactFiberScheduler.js#L928), and after a `setState` to schedule a callback. https://github.com/acdlite/react/blob/867aeac5819c997bfee0a0ddf7d973624ef5e3d2/src/renderers/shared/fiber/ReactFiberScheduler.js#L1475

`PriorityLevel` is also used for going the other direction: priority -> expiration time. When scheduling an update.

File: src/renderers/shared/fiber/ReactFiberScheduler.js (hunk 11 above)
Comment (acdlite) [reply src/renderers/shared/fiber/ReactFiberScheduler.js]: Rather than change the priority context every time we start on a new bucket, and then change it to task in the commit phase, and then reset it to the previous value when an error is thrown, I found it was simpler to move this complexity to `getPriorityContext`. https://github.com/acdlite/react/blob/867aeac5819c997bfee0a0ddf7d973624ef5e3d2/src/renderers/shared/fiber/ReactFiberScheduler.js#L1520-L1552

**Comment by acdlite** [comment]
Addressed nits

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+// TODO: Use an opaque type once ESLint et al support the syntax
+export type ExpirationTime = number;
+
+const NoWork = 0;
+const Sync = 1;
+const Task = 2;
+const Never = Math.pow(2, 31) - 1; // Max int32
+
+const UNIT_SIZE = 10;
+const MAGIC_NUMBER_OFFSET = 10;
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberExpirationTime.js]: Why is this 10 and not 3?

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+const invariant = require('fbjs/lib/invariant');
+
+// TODO: Use an opaque type once ESLint et al support the syntax
+export type ExpirationTime = number;
+
+const NoWork = 0;
+const Sync = 1;
+const Task = 2;
+const Never = Math.pow(2, 31) - 1; // Max int32
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberExpirationTime.js]: Should we just hardcode the value so we don't have to hope compilers do it for us?

File: src/renderers/shared/fiber/ReactFiberUpdateQueue.js [hunk 15]
```diff
@@ -104,6 +83,7 @@ function createUpdateQueue(): UpdateQueue {
 function cloneUpdate(update: Update): Update {
   return {
     priorityLevel: update.priorityLevel,
```
Comment (sebmarkbage) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: If we get rid of coalescing we can get rid of this.

**Comment by acdlite** [comment]
I'll remove coalescing and `PriorityLevel` in a follow-up.

File: src/renderers/shared/ReactDOMFrameScheduling.js [hunk 16]
```diff
@@ -67,19 +81,23 @@ if (!ExecutionEnvironment.canUseDOM) {
...
-          // gets a performance timer value. Not sure if this is always true.
-          return frameDeadline - performance.now();
-        }
-      : function() {
-          // As a fallback we use Date.now.
-          return frameDeadline - Date.now();
-        },
-  };
+  var frameDeadlineObject;
+  if (hasNativePerformanceNow) {
```
Comment (NE-SmallTown) [inline src/renderers/shared/ReactDOMFrameScheduling.js]: Why we do the judgement again instead calling `now()` directly?

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js (hunk 9 above)
Comment (NE-SmallTown) [reply src/renderers/shared/fiber/ReactFiberExpirationTime.js]: `ms / UNIT_SIZE) | 0` equals `Math.floor` rather than `Math.round`, right?

File: src/renderers/shared/fiber/ReactFiberExpirationTime.js (hunk 13 above)
Comment (NE-SmallTown) [reply src/renderers/shared/fiber/ReactFiberExpirationTime.js]: @acdlite Any explanation here?
